public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx

```diff
@@ -0,0 +1,139 @@
+import React, { useState } from 'react';
+import { DataQuery, SelectableValue } from '@grafana/data';
+import { Button, Field, Input, InputControl, Modal, RadioButtonGroup } from '@grafana/ui';
+import { DashboardPicker } from 'app/core/components/editors/DashboardPicker';
+import { FolderPicker } from 'app/core/components/Select/FolderPicker';
+import { DeepMap, FieldError, useForm } from 'react-hook-form';
+import { SaveToExistingDashboardDTO, SaveToNewDashboardDTO } from './addToDashboard';
+
+const isSaveToNewDashboard = (
+  errors: DeepMap<FormDTO, FieldError>,
+  saveTarget: SaveTarget
+): errors is DeepMap<SaveToNewDashboardDTO, FieldError> => saveTarget === 'new';
+
+const options: Array<SelectableValue<SaveTarget>> = [
+  { label: 'New dashboard', value: 'new' },
+  { label: 'Existing dashboard', value: 'existing' },
+];
+
+type SaveTarget = 'new' | 'existing';
+
+type FormDTO = SaveToExistingDashboardDTO | SaveToNewDashboardDTO;
+
+interface Props {
+  onClose: () => void;
+  queries: DataQuery[];
+  visualization: string;
+  onSave: (data: FormDTO, redirect: boolean) => Promise<void | { message: string; status: string }>;
+}
+
+function withRedirect<T extends any[]>(fn: (redirect: boolean, ...args: T) => {}, redirect: boolean) {
+  return async (...args: T) => fn(redirect, ...args);
+}
+
+export const AddToDashboardModal = ({ onClose, queries, visualization, onSave }: Props) => {
+  const [saveTarget, setSaveTarget] = useState<SaveTarget>('new');
+  const {
+    register,
+    handleSubmit,
+    control,
+    formState: { errors, isSubmitting },
+    setError,
+  } = useForm<FormDTO, { error: string }>({ defaultValues: { queries, visualization } });
+
+  const onSubmit = async (withRedirect: boolean, data: FormDTO) => {
+    const error = await onSave(data, withRedirect);
+
+    if (error) {
+      switch (error.status) {
+        case 'name-exists':
+        case 'empty-name':
+        case 'name-match':
+          setError('dashboardName', { message: error.message, shouldFocus: true });
+          break;
+        default:
+        // TODO: Other unknown errors may happen, we should handle them by displaying an error message
```

Comment:
Needs UX: how shall we show errors that are not related to a specific field?

An alert notification (the one appearing on the top-right corner) are not a viable solution as the modal (for accessibility reasons) is going to be the only component readable by screen readers, so the error must be in the modal itself.

cc. @nadinevehling